kbmd/box.c

```diff
@@ -0,0 +1,1057 @@
+	    (ret = envlist_add_string(zcp_args, "dataset",
+	    dataset)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "prop", propstr)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "ebox", eboxstr)) != ERRF_OK)
+		goto done;
+
+	if (!stage) {
+		if ((ret = add_hexkey(zcp_args, "keyhex", key,
+		    keylen)) != ERRF_OK) {
+			goto done;
+		}
+	}
+
+	ret = run_channel_program(dataset, add_prog, zcp_args, &result);
+
+done:
+	mutex_exit(&piv_lock);
+	ebox_free(ebox);
+	free(eboxstr);
+	nvlist_free(zcp_args);
+	nvlist_free(result);
+	freezero(key, keylen);
+
+	if (ret == ERRF_OK) {
+		ret = post_recovery_config_update();
+	}
+
+	return (ret); 
+}
+
+errf_t *
+activate_recovery(const char *dataset)
+{
+	errf_t *ret = ERRF_OK;
+	kbmd_token_t *kt = NULL;
+	struct ebox *ebox = NULL;
+	nvlist_t *zcp_args = NULL;
+	nvlist_t *result = NULL;
+	const uint8_t *key = NULL;
+	size_t keylen = 0;
+
+	mutex_enter(&piv_lock);
+	if (sys_pool == NULL) {
+		mutex_exit(&piv_lock);
+		return (errf("NotFoundError", NULL,
+		    "system zpool must be set before activating a recovery"
+		    "config"));
+	}
+
+	if ((ret = kbmd_get_ebox(dataset, B_TRUE, &ebox)) != ERRF_OK ||
+	    (ret = kbmd_unlock_ebox(ebox, &kt)) != ERRF_OK)
+		goto done;
+
+	key = ebox_key(ebox, &keylen);
+
+	if ((ret = envlist_alloc(&zcp_args)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "dataset",
+	    dataset)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "ebox", BOX_PROP)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "stagedebox",
+	    STAGEBOX_PROP)) != ERRF_OK ||
+	    (ret = add_hexkey(zcp_args, "keyhex", key, keylen)) != ERRF_OK) {
+		goto done;
+	}
+
+	if ((ret = run_channel_program(dataset, activate_prog, zcp_args,
+	    &result)) != ERRF_OK) {
+		goto done;
+	}
+
+	if (strcmp(dataset, sys_pool) == 0) {
+		ebox_free(sys_box);
+		sys_box = ebox;
+		ebox = NULL;
+	}
+
+done:
+	if (kt != sys_piv)
+		kbmd_token_free(kt);
+
+	mutex_exit(&piv_lock);
+	ebox_free(ebox);
+	nvlist_free(zcp_args);
+	nvlist_free(result);
+
+	if (ret == ERRF_OK) {
+		ret = post_recovery_config_update();
+	}
+
+	return (ret);
+}
+
+errf_t *
+remove_recovery(const char *dataset)
+{
+	errf_t *ret = ERRF_OK;
+	zfs_handle_t *zhp = NULL;
+
+	(void) bunyan_trace(tlog, "remove_recovery: enter",
+	    BUNYAN_T_END);
+
+	mutex_enter(&piv_lock);
+
+	if (sys_pool == NULL) {
+		mutex_exit(&piv_lock);
+		return (errf("NotFound", NULL, "system pool is not set"));
+	}
+
+	mutex_enter(&g_zfs_lock);
+
+	if ((ret = ezfs_open(g_zfs, dataset,
+	    ZFS_TYPE_FILESYSTEM|ZFS_TYPE_VOLUME, &zhp)) != ERRF_OK) {
+		mutex_exit(&piv_lock);
+		ret = errf("EBoxError", ret,
+		    "unable to load ebox for %s", sys_pool);
+		goto done;
+	}
+
+	mutex_exit(&piv_lock);
+
+	if ((ret = ezfs_prop_inherit(zhp, STAGEBOX_PROP)) != ERRF_OK) {
+		goto done;
+		/*
+		 * XXX: figure out what is returned when it doesn't exist, and
+		 * convert to non error
+		 */
+	}
```

Comment:
If inherit fails because the property is not set local to that dataset, your work is done.  Claim success and move on.

Reply:
Testing a bit w/ dtrace, inheriting a non-existent user property never returns an error, so I'll remove the XXX comment, and add one indicating that -- if we do get an error, then it's probably something we want to report.

Reply:
confirmed